### ISSUE TITLE
external perl from /usr breaks build dependencies

### ISSUE THREAD

**Issue by mwkrentel** [spack/spack#13792]
One of my cron jobs tripped an alarm this morning.  I've tracked the
problem down to this commit.  (The date says Nov 4, but it was merged
yesterday, Nov 18).

```
commit 99dfff447509a68d672a82812e5abfbc463b6d26
Author: Greg Becker <becker33@llnl.gov>
Date:   Mon Nov 4 23:03:35 2019 -0600

    binary distribution: relocate text files properly in relative binaries (#13578)
    
    * Make relative binaries relocate text files properly
    * rb strings aren't valid in python 2
    * move perl to new interface for setup_environment family methods

M       lib/spack/spack/binary_distribution.py
M       lib/spack/spack/relocate.py
M       var/spack/repos/builtin/packages/perl/package.py
```

I want to build libunwind version 2018.10.12.

```
$ spack spec libunwind @2018.10.12 %gcc@5.3.0
Input spec
--------------------------------
libunwind@2018.10.12%gcc@5.3.0

Concretized
--------------------------------
libunwind@2018.10.12%gcc@5.3.0~xz arch=linux-rhel7-broadwell
    ^autoconf@2.69%gcc@5.3.0 arch=linux-rhel7-broadwell
        ^m4@1.4.18%gcc@5.3.0 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-rhel7-broadwell
            ^libsigsegv@2.12%gcc@5.3.0 arch=linux-rhel7-broadwell
        ^perl@5.16.3%gcc@5.3.0+cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads arch=linux-rhel7-broadwell
    ^automake@1.16.1%gcc@5.3.0 arch=linux-rhel7-broadwell
    ^libtool@2.4.6%gcc@5.3.0 arch=linux-rhel7-broadwell
```

This is on RedHat 7.7, x86_64.  gcc 5.3.0 is from a spack-built module
(/usr/bin/gcc is 4.8.5, hence the module).  This has all worked for
months.

Default config.yaml.  Packages.yaml just has a path system perl.

```
packages:
  perl:
    paths:
      perl@5.16.3:  /usr
    buildable: False
```

It fails with a libtool version problem, something to do with
`/usr/bin/libtool` version 2.4.2 and the spack-built one 2.4.6.

```
/bin/sh ../libtool  --tag=CC   --mode=compile /home/krentel/failure/spack-repo/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I../include  -I../include -I../include/tdep-x86_64 -I. -D_GNU_SOURCE -DNDEBUG  -g -O2 -fexceptions -Wall -Wsign-compare -MT mi/Gget_proc_info_by_ip.lo -MD -MP -MF $depbase.Tpo -c -o mi/Gget_proc_info_by_ip.lo mi/Gget_proc_info_by_ip.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [os-linux.lo] Error 63
make[2]: *** Waiting for unfinished jobs....
```

Note: libunwind 2018.10.12 doesn't include the autotools-generated
files.  So for this version, the autotools prereqs are required to
generate them.  I put this in libuwind `package.py`.

```
depends_on('autoconf', type='build', when='@2018:')
depends_on('automake', type='build', when='@2018:')
depends_on('libtool',  type='build', when='@2018:')
depends_on('m4',       type='build', when='@2018:')
```

Autotools writes the value of PATH in `config.log`.  I tracked this
down to putting `/usr/bin` early in my spack path.

The previous commit has this and works:

```
PATH: /home/krentel/failure/spack-repo/lib/spack/env/gcc
PATH: /home/krentel/failure/spack-repo/lib/spack/env/case-insensitive
PATH: /home/krentel/failure/spack-repo/lib/spack/env
PATH: /home/krentel/failure/install/linux-rhel7-broadwell/gcc-5.3.0/libtool-2.4.6-xvngx4hjmii7ixzx4m23yf2bepp5xwpw/bin
PATH: /home/krentel/failure/install/linux-rhel7-broadwell/gcc-5.3.0/m4-1.4.18-eo2dulb33kd3q5zfw2r2njczfng3wcux/bin
PATH: /home/krentel/failure/install/linux-rhel7-broadwell/gcc-5.3.0/automake-1.16.1-qf74dlj76f2mlzy5jbdkfzedjtzlekzd/bin
PATH: /home/krentel/failure/install/linux-rhel7-broadwell/gcc-5.3.0/m4-1.4.18-eo2dulb33kd3q5zfw2r2njczfng3wcux/bin
...
```

The 99dfff44750 commit has this:

```
PATH: /usr/bin
PATH: /home/krentel/failure/spack-repo/lib/spack/env/gcc
PATH: /home/krentel/failure/spack-repo/lib/spack/env/case-insensitive
PATH: /home/krentel/failure/spack-repo/lib/spack/env
PATH: /home/krentel/failure/install/linux-rhel7-broadwell/gcc-5.3.0/libtool-2.4.6-xvngx4hjmii7ixzx4m23yf2bepp5xwpw/bin
PATH: /home/krentel/failure/install/linux-rhel7-broadwell/gcc-5.3.0/m4-1.4.18-eo2dulb33kd3q5zfw2r2njczfng3wcux/bin
PATH: /home/krentel/failure/install/linux-rhel7-broadwell/gcc-5.3.0/automake-1.16.1-qf74dlj76f2mlzy5jbdkfzedjtzlekzd/bin
...
```

This is going to break a lot of build dependencies if you put
`/usr/bin` at the front of PATH.

ping @becker33

**Comment by mwkrentel** [comment]
@tgamblin  @becker33 

I looked a little deeper into this.  It's mostly just a problem with
perl.  Something from the 99dfff44 commit is putting the path to perl
at the FRONT of PATH.

So, whenever you have a perl dependency with `packages.yaml`
specifying perl from /usr/bin, this will pretty much break all build
dependencies.  With /usr/bin at the front of PATH, this will cover up
the spack-built programs.

There are two workarounds:

1. Use e433a5dc0b, the commit before 99dfff44.

2. Have spack always build perl, don't use one from a system
directory.

But this brings up a larger problem.  For any external dependency
listed in `packages.yaml` from /usr, you better put /usr/bin after all
the spack-built paths (and -I and -L at the end of flags), or else
you'll hide the spack-built prereq.

**Comment by mwkrentel** [comment]
Here's another example to illustrate the severity of the breakage.

I tried building mbedtls on RH 7.7.  Mbedtls is a small cmake package,
but it needs a cmake newer than the 2.8.12 that comes with RH 7.x.
(`package.py` says 2.6, but that's old and a separate issue.)
The build runs a couple of perl scripts, so it also depends on perl.

I use this for `packages.yaml`, perl from `/usr/bin` and a local build
of cmake to get a later version.

```
packages:
  cmake:
    paths:
      cmake@3.11.4:  /projects/pkgs/cmake
   buildable: False

  perl:
    paths:
      perl@5.16.3:  /usr
    buildable: False
```

`spack-build-env.txt` shows the value of the build PATH (with lines
folded and ':' translated to ' ').

```
PATH=/usr/bin
/home/krentel/failure/spack-repo/lib/spack/env/gcc
/home/krentel/failure/spack-repo/lib/spack/env/case-insensitive
/home/krentel/failure/spack-repo/lib/spack/env
/projects/pkgs/cmake/bin
/projects/pkgs/cmake/bin
/home/krentel/Modules/install/linux-rhel7-x86_64/gcc-4.8.5/gcc-7.3.0-qemsqrcwkk52f6neef4kg5wvoucsroif/bin
/home/krentel/viewers/install/bin/
/projects/pkgs/cmake/bin/
/projects/pkgs/autotools/bin
...
```

Again, `/usr/bin` is at the front of PATH, so the cmake becomes
`/usr/bin/cmake` instead of the one specified in `packages.yaml`.
And indeed the build fails with:

```
     18    -- Found Threads: TRUE
  >> 19    CMake Error at programs/ssl/CMakeLists.txt:37 (target_sources):
     20      Unknown CMake command "target_sources".
```

In summary, any package that has a direct build dependency on perl or
an indirect run dependency on perl is at risk for breaking ALL of its
other build dependencies.

For many things like sed, awk, grep, tar, you won't notice because
these are not version sensitive.  But you'll be running the `/usr/bin`
version, not what spack `depends_on` says.  And other build tools like
cmake, python, libtool are more version sensitive.

**Comment by mwkrentel** [comment]
@scheibelp I saw your fix for this in #13903.  Thanks!

But I was a little concerned at the label 'impact-low'.  That comes
across like spack doesn't really want to support the case of using
perl from /usr/bin and instead you want everyone to rebuild perl,
whether they need it or not.  Is that correct?

For my use case (hpctoolkit), perl is just a build tool, we don't use
it at run time, /usr/bin/perl almost always works, and I'd prefer not
to rebuild it when I don't need to.
